crates/types/src/net/mod.rs

```diff
@@ -51,22 +52,57 @@ pub enum AdvertisedAddress {
     /// Unix domain socket
     #[display("unix:{}", _0.display())]
     Uds(PathBuf),
-    /// Hostname or host:port pair, or any unrecognizable string.
+    /// Hostname or host:port pair
     #[display("{}", _0)]
     Http(Uri),
 }
 
 impl FromStr for AdvertisedAddress {
-    type Err = http::uri::InvalidUri;
+    type Err = Error;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        if let Some(stripped_address) = s.strip_prefix("unix:") {
-            Ok(AdvertisedAddress::Uds(
-                stripped_address.parse().expect("infallible"),
-            ))
-        } else {
-            // try to parse as a URI
-            Ok(AdvertisedAddress::Http(s.parse()?))
+        match s.trim() {
+            "" => Err(anyhow::anyhow!("Advertised address cannot be empty")),
+            address if address.starts_with("unix:") => {
+                parse_uds(&address[5..]).map(AdvertisedAddress::Uds)
+            }
+            address => parse_http(address),
+        }
+    }
+}
+
+fn parse_uds(s: &str) -> Result<PathBuf, Error> {
+    s.parse::<PathBuf>()
+        .with_context(|| format!("Failed to parse Unix domain socket path: '{}'", s))
+}
+
+fn parse_http(s: &str) -> Result<AdvertisedAddress, Error> {
+    let uri = s
+        .parse::<Uri>()
+        .with_context(|| format!("Invalid URI format: '{}'", s))?;
+
+    match uri.scheme_str() {
+        Some("http") | Some("https") => Ok(AdvertisedAddress::Http(uri)),
+        Some(other) => Err(anyhow::anyhow!("Unsupported URI scheme '{}'", other)),
+        None => Err(anyhow::anyhow!("Missing URI scheme in: '{}'", s)),
+    }
+}
+
+impl AdvertisedAddress {
+    /// Derives a `BindAddress` based on the advertised address.
+    pub fn derive_bind_address(&self) -> Option<BindAddress> {
+        match self {
+            AdvertisedAddress::Http(uri) => {
+                uri.authority()
+                    .and_then(|auth| auth.port_u16())
+                    .map(|port| {
+                        BindAddress::Socket(SocketAddr::new(
+                            IpAddr::V4(Ipv4Addr::UNSPECIFIED),
```

Comment:
Any chance we can leave the door open for IPv6 without over-specifying IPv4 here?

Reply:
updated 🔢